File: 2023/Day_01/fSharp/Program.fs
open System
open System.IO

// Setup
let substitutes =
    Map
        [ ("one", 1)
          ("1", 1)
          ("two", 2)
          ("2", 2)
          ("three", 3)
          ("3", 3)
          ("four", 4)
          ("4", 4)
          ("five", 5)
          ("5", 5)
          ("six", 6)
          ("6", 6)
          ("seven", 7)
          ("7", 7)
          ("eight", 8)
          ("8", 8)
          ("nine", 9)
          ("9", 9) ]

let input = File.ReadAllLines "../src/input.txt" |> List.ofArray

// Util
let isDigit (c: char) = Char.IsDigit c
let toNumber (c: char) = int c - int '0'
let toResult (i: int, j: int) = i * 10 + j
let together f g x = (f x, g x)

// TASK 1
let get findFn (s: string) = findFn isDigit s |> toNumber

let getCalibration1 (s: string) =
    together (get Seq.find) (get Seq.findBack) s |> toResult

// Task 1 Result
List.map getCalibration1 input |> List.sum |> printf "%i\n"

// TASK 2
let findKey (baseString: string) (searchString: string) =
    (baseString.IndexOf searchString, searchString)

let findLastKey (baseString: string) (searchString: string) =
    (baseString.LastIndexOf searchString, searchString)

let get2 indexFn orderingFn (s: string) =
    Seq.map (indexFn s) substitutes.Keys
    |> Seq.filter (fun (index, digitWord) -> index <> -1)
    |> orderingFn fst
    |> fun x -> substitutes[snd x]

let getFirst2 (s: string) = get2 findKey Seq.minBy s
let getLast2 (s: string) = get2 findLastKey Seq.maxBy s

let getCalibration2 (x: string) =
    together getFirst2 getLast2 x |> toResult

// Task 2 Result
List.map getCalibration2 input |> List.sum |> printf "%i\n"

File: 2023/Day_01/src/Program.cs
namespace Day_01;
public static class Program
{
    static void Main(string[] args)
    {
        var input = File.ReadLines("input.txt").ToList();
        Console.WriteLine(Task1(input));
        Console.WriteLine(Task2(input));
    }

    public static int Task1(List<string> input) => input.Sum(GetDigits);
    public static int GetDigits(string line) => (line.First(Char.IsDigit) - '0') * 10 + (line.Last(Char.IsDigit) - '0');

    public static int Task2(List<string> input) => input.Sum(GetDigits2);
    public static int GetDigits2(string line)
    {
        (int Index, int Value) first = (99, 0);
        (int Index, int Value) last = (-2, 0);

        foreach (var digit in DigitWords.Keys)
        {
            var index = line.IndexOf(digit);
            if (index != -1 && index < first.Index)
                first = (index, DigitWords[digit]);
            var indexLast = line.LastIndexOf(digit);
            if (indexLast != -1 && indexLast > last.Index)
                last = (indexLast, DigitWords[digit]);
        }
        return first.Value * 10 + last.Value;
    }

    static readonly Dictionary<string, int> DigitWords = new(){
        { "1", 1},        { "one",   1},
        { "2", 2},        { "two",   2},
        { "3", 3},        { "three", 3},
        { "4", 4},        { "four",  4},
        { "5", 5},        { "five",  5},
        { "6", 6},        { "six",   6},
        { "7", 7},        { "seven", 7},
        { "8", 8},        { "eight", 8},
        { "9", 9},        { "nine",  9},
    };
}
